src/Key.php

```diff
@@ -5,18 +5,132 @@
 class Key
 {
     /** @var string */
-    public $id;
+    private $id;
 
     /** @var string */
-    public $secret;
+    private $secret;
+
+    /** @var string */
+    private $certificate;
+
+    /** @var string */
+    private $privateKey;
+
+    /** @var string */
+    private $type;
 
     /**
      * @param string $id
      * @param string $secret
      */
-    public function __construct($id, $secret)
+    public function __construct($id, $item)
     {
         $this->id = $id;
-        $this->secret = $secret;
+        $certificate = $this->getX509Certificate($item);
+        $privateKey = $this->getRSAPrivateKey($item);
+        if (($certificate || $privateKey)) {
+            $this->type = 'rsa';
+            if ($privateKey) {
+                $this->privateKey = $privateKey;
+            }
+            if ($certificate) {
+                $this->certificate = openssl_x509_read($certificate);
+            }
+            if ($certificate && $privateKey) {
+                if (!openssl_x509_check_private_key(
+                        $this->certificate,
+                        $this->privateKey
+                    )
+                ) {
+                    throw new KeyException('Supplied Certificate and Key are not related');
+                }
+            }
+        } else {
+            $this->type = 'secret';
+            $this->secret = $item;
+            $publicKey = null;
+            $privateKey = null;
+        }
+    }
+
+    private function getRSAPrivateKey($object)
+    {
+        if (is_array($object)) {
+            foreach ($object as $item) {
+                $privateKey = Key::getRSAPrivateKey($item);
+                if ($privateKey) {
+                    return $privateKey;
+                }
+            }
+        } else {
+            try {
+                $privateKey = openssl_get_privatekey($object);
+            } catch (\Exception $e) {
+                $privateKey = null;
+            }
+            if ($privateKey) {
+                return $privateKey;
+            }
+        }
+    }
+
+    private function getX509Certificate($object)
+    {
+        $key = null;
+        if (is_array($object)) {
+            foreach ($object as $item) {
+                $result = Key::getX509Certificate($item);
+                if ($result) {
+                    $key = $result;
+                }
+            }
+
+            return $key;
+        } else {
+            try {
+                $result = openssl_get_publickey($object);
+            } catch (\Exception $e) {
+                $result = null;
+            }
```

Comment:
why are you try / catching ?

Reply:
We're not specifying which order the public/private keys should be provided, and no specific Key subclasses for each, so have to detect:
- If the result is not a public key, ignore it - the same object could resolve to a private key which is useless for verification.
- Same for openssl_get_privatekey() test.

Will see if there's a non-exception based detection mechanism, but the openssl libraries aren't very friendly.